src/main/mondrian/olap/fun/CrossJoinFunDef.java

```diff
@@ -565,7 +567,10 @@ private static void cartesianProductRecurse(
         final int partialSizeNext = partialSize + tupleList.getArity();
         final int iNext = i + 1;
         final TupleCursor cursor = tupleList.tupleCursor();
+        int currentIteration = 0;
         while (cursor.forward()) {
+            CancellationChecker.checkCancelOrTimeout(
+                currentIteration++, Locus.peek().execution);
```

Comment:
Can you put Execution in a variable please? It will always be the same execution and will remove a few virtual function calls from each iteration, as this is a tight loop. (please apply this in all other instances)